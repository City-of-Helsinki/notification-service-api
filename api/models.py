# Create your models here.
from copy import deepcopy

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from django.contrib.postgres.fields import JSONField
from django.db import models
from django.utils.translation import ugettext_lazy as _


class DeliveryLog(UUIDPrimaryKeyModel, TimestampedModel):
    user = models.ForeignKey(
        "users.User",
        related_name="delivery_logs",
        on_delete=models.CASCADE,
        verbose_name=_("user"),
    )
    report = JSONField(verbose_name=_("report"), blank=True, null=True)

    class Meta:
        verbose_name = _("delivery log")
        verbose_name_plural = _("delivery logs")
        ordering = ["-updated_at"]

    def update_report(self, report_data):
        """
        Update delivery of single message base on json data sent from Quriiri to
        our service
        :param report_data: JSON
        :return:
        """
        report = self.report
        updated_messages = deepcopy(report["messages"])
        destination = report_data["destination"]
        for k, v in report["messages"].items():
            if k == destination or v["converted"] == destination:
                updated_messages[k] = report_data
        report["messages"] = updated_messages
        self.report = report
        self.save()
