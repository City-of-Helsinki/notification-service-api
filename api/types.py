from typing import List, Optional, TypedDict


class Recipient(TypedDict):
    destination: str
    format: Optional[str]


class SendMessagePayload(TypedDict):
    sender: str
    to: List[Recipient]
    text: str


class MessageWebhookPayload(TypedDict):
    sender: str
    destination: str
    status: str
    statustime: str
    smscount: str
    billingref: str
