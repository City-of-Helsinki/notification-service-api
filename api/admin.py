from api.models import DeliveryLog
from django.contrib import admin

admin.site.register(DeliveryLog)
