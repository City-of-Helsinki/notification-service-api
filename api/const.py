NOTIFICATION_TYPE_MOBILE = "mobile"
REGION = "FI"
