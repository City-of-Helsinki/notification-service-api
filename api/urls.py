from django.urls import path

from . import views

urlpatterns = [
    path("message/send", views.send_message, name="send_message"),
    path("message/<id>", views.get_delivery_log, name="get_message"),
    path(
        "message/webhook/<id>",
        views.delivery_log_webhook,
        name="delivery_log_webhook",
    ),
]
