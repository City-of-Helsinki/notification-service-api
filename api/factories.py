import factory
from api.models import DeliveryLog
from users.factories import ApiUserFactory


class DeliveryLogFactory(factory.django.DjangoModelFactory):
    payload = factory.Faker("pystr", max_chars=255)
    api_user = factory.SubFactory(ApiUserFactory)

    class Meta:
        model = DeliveryLog
