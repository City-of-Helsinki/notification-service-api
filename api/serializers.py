from rest_framework import serializers

from api.models import DeliveryLog


class DeliveryLogSerializer(serializers.ModelSerializer):
    report = serializers.JSONField()

    class Meta:
        model = DeliveryLog
        fields = ("id", "report")
