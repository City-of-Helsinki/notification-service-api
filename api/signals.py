from django.conf import settings
from django.db.models.signals import post_delete, post_init, post_save
from django.dispatch import receiver
from rest_framework.authtoken.models import Token

from api.models import DeliveryLog
from audit_log.enums import Operation, Status
from audit_log.services import audit_log_service
from audit_log.types import AuditCommitMessage
from audit_log.utils import create_commit_message


@receiver(post_save, sender=settings.AUTH_USER_MODEL)
def create_auth_token(sender, instance=None, created=False, **kwargs):
    if created:
        Token.objects.create(user=instance)


@receiver(post_save, sender=DeliveryLog)
def audit_log_delivery_log_create_update(
    sender, instance, created, path="", ip_address="", **kwargs
):
    message = create_commit_message(
        status=Status.SUCCESS.value,
        operation=Operation.CREATE.value if created else Operation.UPDATE.value,
        actor=audit_log_service._get_actor_data(
            user=instance.user, ip_address=ip_address
        ),
        target=audit_log_service._get_target(path=path, object_ids=[str(instance.id)]),
    )
    audit_log_service._commit_to_audit_log(message=AuditCommitMessage(**message))


@receiver(post_delete, sender=DeliveryLog)
def audit_log_delivery_log_delete(sender, instance, path="", ip_address="", **kwargs):
    message = create_commit_message(
        status=Status.SUCCESS.value,
        operation=Operation.DELETE.value,
        actor=audit_log_service._get_actor_data(
            user=instance.user, ip_address=ip_address
        ),
        target=audit_log_service._get_target(path=path, object_ids=[str(instance.id)]),
    )
    audit_log_service._commit_to_audit_log(message=AuditCommitMessage(**message))


@receiver(post_init, sender=DeliveryLog)
def audit_log_delivery_log_read(sender, instance, path="", ip_address="", **kwargs):
    message = create_commit_message(
        status=Status.SUCCESS.value,
        operation=Operation.READ.value,
        actor=audit_log_service._get_actor_data(
            user=instance.user, ip_address=ip_address
        ),
        target=audit_log_service._get_target(path=path, object_ids=[str(instance.id)]),
    )
    audit_log_service._commit_to_audit_log(message=AuditCommitMessage(**message))
