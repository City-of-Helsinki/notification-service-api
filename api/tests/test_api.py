import uuid

import pytest
from django.urls import reverse
from rest_framework.authtoken.models import Token

from api.factories import DeliveryLogFactory
from api.models import DeliveryLog
from api.types import MessageWebhookPayload, SendMessagePayload
from common.tests.mock_data import QURIIRI_SMS_RESPONSE


@pytest.fixture(autouse=True)
def autouse_db(db):
    """Rename the db fixture"""
    pass


SMS_PAYLOAD: SendMessagePayload = {
    "sender": "Hel.fi",
    "to": [{"destination": "+358461231231", "format": "MOBILE"}],
    "text": "SMS message",
}

SMS_WEBHOOK_DATA: MessageWebhookPayload = {
    "sender": "hel.fi",
    "destination": "+358461231231",
    "status": "DELIVERED",
    "statustime": "2020-07-21T09:18:00Z",
    "smscount": "1",
    "billingref": "Palvelutarjotin",
}


def test_send_sms_unauthentication(anonymous_api_client):
    response = anonymous_api_client.post(
        reverse("send_message"), SMS_PAYLOAD, format="json"
    )
    assert response.status_code == 401


def test_send_sms(token_api_client, snapshot, mock_send_sms):
    response = token_api_client.post(
        reverse("send_message"), SMS_PAYLOAD, format="json"
    )
    assert DeliveryLog.objects.count() == 1
    snapshot.assert_match(response.data["report"])


def test_send_sms_bad_request(token_api_client, snapshot, mock_send_sms):
    response = token_api_client.post(
        reverse("send_message"),
        {
            "sender": "Hel.fi",
            "text": "Missing the 'to' from data.",
        },
        format="json",
    )
    assert response.status_code == 400
    assert DeliveryLog.objects.count() == 0
    assert (
        response.content.decode()
        == "Missing required keys: 'sender', 'to', and 'text' are required."
    )  # Check the error message


def test_webhook_delivery_log(
    token_api_client, anonymous_api_client, snapshot, mock_send_sms
):
    response = anonymous_api_client.post(
        reverse("delivery_log_webhook", kwargs={"id": uuid.uuid4()})
    )
    assert response.status_code == 404

    # Send a message to initiate delivery log
    token_api_client.post(reverse("send_message"), SMS_PAYLOAD, format="json")

    assert DeliveryLog.objects.count() == 1
    log = DeliveryLog.objects.first()
    snapshot.assert_match(log.report)

    webhook_data = SMS_WEBHOOK_DATA

    anonymous_api_client.post(
        reverse("delivery_log_webhook", kwargs={"id": str(log.id)}),
        data=webhook_data,
        format="json",
    )
    log.refresh_from_db()
    snapshot.assert_match(log.report)


def test_get_delivery_log_unauthenticated_or_unauthorized(
    anonymous_api_client, token_api_client
):
    log = DeliveryLogFactory()
    response = anonymous_api_client.get(
        reverse("get_message", kwargs={"id": str(log.id)})
    )
    # Unauthenticated
    assert response.status_code == 401

    response = token_api_client.get(reverse("get_message", kwargs={"id": str(log.id)}))
    # Not found
    assert response.status_code == 404


def test_get_delivery_log(token_api_client, snapshot):
    log = DeliveryLogFactory(report=QURIIRI_SMS_RESPONSE)
    token, _ = Token.objects.get_or_create(user=log.user)
    token_api_client.credentials(HTTP_AUTHORIZATION="Token " + token.key)
    response = token_api_client.get(reverse("get_message", kwargs={"id": str(log.id)}))
    snapshot.assert_match(response.data["report"])
