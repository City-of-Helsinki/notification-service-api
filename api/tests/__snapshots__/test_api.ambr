# serializer version: 1
# name: test_get_delivery_log
  dict({
    'errors': list([
    ]),
    'messages': dict({
      '+358461231231': dict({
        'converted': '+358461231231',
        'status': 'CREATED',
      }),
    }),
    'warnings': list([
    ]),
  })
# ---
# name: test_send_sms
  dict({
    'errors': list([
    ]),
    'messages': dict({
      '+358461231231': dict({
        'converted': '+358461231231',
        'status': 'CREATED',
      }),
    }),
    'warnings': list([
    ]),
  })
# ---
# name: test_webhook_delivery_log
  dict({
    'errors': list([
    ]),
    'messages': dict({
      '+358461231231': dict({
        'converted': '+358461231231',
        'status': 'CREATED',
      }),
    }),
    'warnings': list([
    ]),
  })
# ---
# name: test_webhook_delivery_log.1
  dict({
    'errors': list([
    ]),
    'messages': dict({
      '+358461231231': dict({
        'billingref': 'Palvelutarjotin',
        'destination': '+358461231231',
        'sender': 'hel.fi',
        'smscount': '1',
        'status': 'DELIVERED',
        'statustime': '2020-07-21T09:18:00Z',
      }),
    }),
    'warnings': list([
    ]),
  })
# ---
